Erc20 guide (#214)
casper-network/docs: source/docs/casper/workflow/erc20-token-deployment-guide.md

# Deploying an ERC-20 Compliant Token on Casper

On the Casper Network, it is possible to create smart contracts that emulate ERC-20 tokens on Ethereum. These tokens have all of the capabilities of traditional ERC-20 tokens, allowing one to approve, transfer, inquire the balance of, etc. By following this guide, you'll be able to create your own ERC-20 token and deploy it to the Casper Network.

## Prerequisites

- Set up your machine as per the [prerequisites](https://casper.network/docs/workflow/setup).

- [Install the `casper-client`](https://casper.network/docs/workflow/setup).
- [Set up an account](https://casper.network/docs/workflow/setup#setting-up-an-account).
- [Fund your account](https://casper.network/docs/workflow/setup#fund-your-account).

- You will need to have [Node.js](https://nodejs.org/en/) installed. Follow the [instructions](https://nodejs.org/en/download) to install it on your system.

- Lastly, you'll need at least a basic understanding of the [bash command line](https://www.gnu.org/software/bash/manual/bash.html). *`zsh` and other related shells should work as well.*

## Cloning the JavaScript Interface

> For this tutorial, we'll be starting in the home (`~`) directory. Navigate here first:

Reviewer: ```suggestion
We'll be starting in the home (`~`) directory for this tutorial. Navigate here first:
```

`git clone https://github.com/casper-ecosystem/casper-erc20-js-interface.git`

Change your working directory to the project directory:

`cd casper-erc20-js-interface`

Install the required dependencies:

`npm install`

## Edit the Deployment Script

Start by opening the file `index.js` in your preferred text editor or IDE.

We'll start by redefining our constants. These start on line `5` and should look like the following:

```javascript
const NAME = "Test Token";
const SYMBOL = "TST";
const PRECISION = 8;
const TOTAL_SUPPLY = 1_000_000_000;
const GAS_LIMIT = 60_000_000_000; //motes
const WASM_PATH = "./erc20_token.wasm";
const NODE_ADDRESS = "http://162.55.132.188:7777/rpc";
const NETWORK_NAME = "casper-test";
```

Let's take a look at what these constants refer to.

* `NAME`: The name of the ERC-20 contract.
* `SYMBOL`: The symbol of the ERC-20 contract.
* `PRECISION`: The number of decimal places the token can be fractionalized to.
* `TOTAL_SUPPLY`: The total supply of your ERC-20 token.
* `GAS_LIMIT`: The gas limit in motes that will be used to pay for the deployment.
* `WASM_PATH`: The path to the compiled contract.

> * `NODE_ADDRESS`: The validator node used to submit the deploy. The address listed directs to a valid, online node, but this may change in the future. If this node does not respond, you can select another online peer from the list [here](https://testnet.cspr.live/tools/peers). Note that you'll need to replace the port with `7777` for most nodes, as well as will need to add `/rpc` to the end of the address.

Reviewer: ```suggestion
* `NODE_ADDRESS`: The validator node used to submit the deploy. The address listed directs to a valid online node, but this may change in the future. If this node does not respond, you can select another online peer from the list [here](https://testnet.cspr.live/tools/peers). Note that you'll need to replace the port with `7777` for most nodes and add `/rpc` to the end of the address.
```

* `NETWORK_NAME`: The name of the network to which you'll be deploying. By default, we have the Casper Testnet specified as `"casper-test"`. To deploy on the Mainnet, you may change this to `"casper"`.

As long as you generated the keys with the aforementioned command within your project's root folder, the paths to your keys should be the same as already written in the code. Otherwise, you'll need to put in the alternate path to your keys in the `KEYS` constant.

```javascript
const KEYS = Keys.Ed25519.parseKeyFiles(
  "./keys/public_key.pem",
  "./keys/secret_key.pem"
);
```

## Install the Contract

To install the contract, execute the following command:

`npm run erc20iface deploy`

You'll know your deployment succeeded when the name of the contract is printed out in the console.

`... Contract name: Test Token`

> Your ERC-20 token contract is now successfully installed. Next we'll go about transferring tokens.

Reviewer: ```suggestion
Your ERC-20 token contract is now successfully installed. Next, we'll go about transferring tokens.
```

**Note:**

An error message will be provided if a deploy fails. Deployments may fail for a number of reasons, including out-of-gas errors, and internet connectivity issues.

*For help, run `npm run erc20iface help`*

## Transfer Tokens

You can send ERC-20 compliant tokens on the Casper Network as per the ERC specification. We'll use the same  `erc20iface` script to execute the transfer deployments.

> You will need an amount and a destination to execute a transferral. The amount corresponds to the number of tokens you want transferred, and the destination is the hexadecimal public key of the receiving account. Your command should look like this

Reviewer: ```suggestion
You will need an amount and a destination to execute a transfer. The amount corresponds to the number of tokens you want to transfer, and the destination is the hexadecimal public key of the receiving account. Your command should look like this:
```

`npm run erc20iface transfer 200 0166795bb8895dcec5631690fa3d5dd3daacde7efeefb1e79176e9d879fd669b47`

To send tokens from the address you just funded, just change the `KEYS` constant to use the path of that account's public and secret keys (in this case account `0166795bb8895dcec5631690fa3d5dd3daacde7efeefb1e79176e9d879fd669b47`).

## Conclusion

> Thank you for reading Casper's ERC-20 deployment guide. Next, you can learn how to write your own [ERC-20 smart contracts](https://casper.network/docs/erc20), and use [Casper's JavaScript SDKs](https://casper.network/docs/dapp-dev-guide/sdk/script-sdk).

Reviewer: ```suggestion
Thank you for reading Casper's ERC-20 deployment guide. Next, you can learn how to write your own [ERC-20 smart contracts](https://casper.network/docs/erc20), and use [Casper's JavaScript SDK](https://casper.network/docs/dapp-dev-guide/sdk/script-sdk).
```

> Thank you for reading Casper's ERC-20 deployment guide. Learn how to write your own ERC-20 smart contracts by following [this guide](https://casper.network/docs/erc20). To learn more about Casper's JavaScript SDKs read [this guide](https://casper.network/docs/dapp-dev-guide/sdk/script-sdk).

Reviewer: ```suggestion
Thank you for reading Casper's ERC-20 deployment guide. Next, you can learn how to write your own [ERC-20 smart contracts](https://casper.network/docs/erc20), and use [Casper's JavaScript SDKs](https://casper.network/docs/dapp-dev-guide/sdk/script-sdk).
```

# Deploy an ERC-20 Compliant Token to the Casper Network with Node.js

On the Casper Network, it is possible to create smart contracts that emulate ERC-20 tokens on Ethereum. These tokens have all of the capabilities of traditional ERC-20 tokens, allowing one to approve, transfer, inquire the balance of, etc. By following this guide, you'll be able to create your own ERC-20 token and deploy it to the Casper Network.

## Prerequisites

> 

Reviewer 2: We usually put this section in point form and shortened form - refer to link - http://localhost:3000/docs/dapp-dev-guide/tutorials/erc20/deploy#pre-requisites

Author: Changed this to bullet-list.

> Before following this guide ensure that you are familiar with [compiling smart contracts](https://casper.network/docs/dapp-dev-guide/getting-started) and [deploying smart contracts](https://casper.network/docs/dapp-dev-guide/deploying-contracts) on the Casper Network. You will also need to be familiar with the [Casper Signer](https://chrome.google.com/webstore/detail/casperlabs-signer/djhndpllfiibmcdbnmaaahkhchcoijce?hl=en) browser extension.

Author: Ended up not needing the Signer for the guide at all.

You will also need to have [Node.js](https://nodejs.org/en/) installed. Follow the [instructions](https://nodejs.org/en/download) to install it on your system.

Lastly, you'll need at least a basic understanding of the [bash command line](https://www.gnu.org/software/bash/manual/bash.html). *`zsh` and other related shells should work as well.*

## Clone ERC-20 JavaScript Interface

For this tutorial, we'll be starting in the home (`~`) directory. Navigate here first:

`cd ~`

> Now clone the prewritten JavaScript deployer project, so you don't have to reinvent the wheel. This project includes a precompiled ERC-20 contract with only basic functionality. If you'd like to write your own ERC-20 contract with custom logic, please follow the instructions [here](https://casper.network/docs/writing-contracts). Execute this command in your home directory:

Reviewer 2: I would suggest adding the hyperlink to the project with these words - JavaScript deployer project

`git clone https://github.com/casper-ecosystem/casper-erc20-js-interface.git`

Change your working directory to the project directory:

`cd casper-erc20-js-interface`

> Install the required dependencies

Reviewer 2: Missing :

Author: Added 👍

`npm install`

> ## Set up an Account

Reviewer: I recommend making this entire section + funding the account a prerequisite, since we already have a guide for Account setup, funding the account, and casper-client setup under the How To's: https://casper.network/docs/workflow/setup

### Install Rust

To interact with the necessary Rust packages we will first need to install Rust. To do so, execute the following command:

`curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh`

You can also [install Rust using Homebrew](https://formulae.brew.sh/formula/rust) on macOS or `apt` on Debian based linux systems.

### Install `casper-client`

> We will be using `casper-client` to generate a keypair that represents an account on Casper.

Author: Yes, my mistake

> To install the `casper-client` execute the command

Reviewer 2: Missing : at the end of the sentence

`cargo install casper-client`

*`cargo` is a Rust package manager and is installed alongside Rust. Executing the above command installs `casper-client` globally on your system.*

### Generate Account Keys

In the working directory of your project, generate a keypair by running:

`casper-client keygen keys/`

This will generate keys and write them to the `keys` directory. If `keys` does not exist, `casper-client` will create the directory.

**Note:**

You can also create an account with [Casper Signer](https://chrome.google.com/webstore/detail/casperlabs-signer/djhndpllfiibmcdbnmaaahkhchcoijce), or use an account you've made previously. (Note that new versions of Casper Signer do not download the necessary public key when exporting accounts to files).

> 

Reviewer 3: Can we instead say "...the new versions of Casper Signer only download the secret key when exporting accounts to files."

Reviewer 3: Also, there is a note in brackets inside an overall **Note**. Can we remove the brackets and "Note that" text?

> ## Fund the Account with Test CSPR

Reviewer: Similar to my previous comment, this is already covered in the prerequisites of the How To's: https://casper.network/docs/workflow/setup

To fund the test account, you'll first need to import your account in the Casper Signer.

To do this, open the Signer in your browser, unlock your vault with your password ([or set up a new vault](https://casper.network/docs/workflow/signer-guide#12-logging-in-to-the-casper-signer)).

> Click **Import Account**. Name the imported account and click **Upload**. Select the *secret_key.pem* file in the aforementioned *keys* directory.

Reviewer 3: Might look better as numbered steps?
1. Click **Import Account**. 
2. Name the imported account and click **Upload**. 
3. Select the *secret_key.pem* file in the aforementioned *keys* directory.

Author: Agreed, changed.

Import the account and select it as your current active account.

Visit the [testnet faucet](https://testnet.cspr.live/tools/faucet), sign in with Casper Signer, complete the captcha, and request your tokens.

> After a moment, you're account will be credited 1000 CSPR.

Reviewer 2: 'you're' should be 'your', right?

> After a moment, you're account will be credited 1000 CSPR.

Reviewer 3: Consider this "After a moment, 1000 CSPR is credited to your account."

**Note:**

*You can only request tokens once per account.*

## Edit the Deployment Script

Start by opening the file `index.js` in your preferred text editor or IDE.

We'll start by redefining our constants. These start on line `5` and should look like the following.

```javascript
const NAME = "Test Token";
const SYMBOL = "TST";
const PRECISION = 8;
const TOTAL_SUPPLY = 1_000_000_000;
const GAS_LIMIT = 60_000_000_000; //motes
const WASM_PATH = "./erc20_token.wasm";
const NODE_ADDRESS = "http://162.55.132.188:7777/rpc";
const NETWORK_NAME = "casper-test";
```

Let's take a look at what these constants refer to.

* `NAME`: The name of the ERC-20 contract.
* `SYMBOL`: The symbol of the ERC-20 contract.
* `PRECISION`: The number of decimal places the token can be fractionalized to.
* `TOTAL_SUPPLY`: The total supply of your ERC-20 token.
* `GAS_LIMIT`: The gas limit in motes that will be used to pay for the deployment
* `WASM_PATH`: The path to the compiled contract.

> * `NODE_ADDRESS`: The address of the validator node to submit the deployment to. The address listed directs to a valid, online node, but this may change in the future. If this node does not respond, you can select another online peer from the list [here](https://testnet.cspr.live/tools/peers). Note that you'll need to replace the port with `7777` for most nodes, as well as will need to add `/rpc` to the end of the address.

Reviewer 3: Could we instead say "The validator node used to submit the deploy"?

> * `NETWORK_NAME`: The name of the network you'll be deploying to. By default we have the casper testnet chosen with `"casper-test"`. To deploy on the mainnet, you may change this to `"casper"`.

Reviewer: ```suggestion
* `NETWORK_NAME`: The name of the network to which you'll be deploying. By default, we have the Casper Testnet specified as `"casper-test"`. To deploy on the Mainnet, you may change this to `"casper"`.
```

As long as you generated the keys with the aforementioned command within your project's root folder, the paths to your keys should be the same as already written in the code. Otherwise you'll need to put in the alternate path to your keys here, on line `14`.

> 

Reviewer 3: Consider "Otherwise, you'll need to put in the alternate path to your keys in the `KEYS` constant."

```javascript
const KEYS = Keys.Ed25519.parseKeyFiles(
  "./keys/public_key.pem",
  "./keys/secret_key.pem"
);
```

## Install the Contract

To install the contract, execute the following command:

`npm run erc20iface deploy`

> If deployment succeeds, your last log line will tell you the name of the installed token.

Reviewer: It would be good to add some details on how to know whether or not the deploy succeeded. Maybe also mention possible reasons a deploy might fail (not enough gas specified).

`... Contract name: Test Token`

Your ERC-20 token contract is now successfully installed. Next we'll go about transferring tokens.

**Note:**

*For help, run `npm run erc20iface help`*

## Transfer Tokens

> Casper Network's ERC-20 tokens can be sent back and forth as per the ERC specification. We'll be using the same  `erc20iface` script to execute the transfer deployments.

Reviewer: ```suggestion
You can send ERC-20 compliant tokens on the Casper Network as per the ERC specification. We'll use the same  `erc20iface` script to execute the transfer deployments.
```

You will need an amount and a destination to execute a transferral. The amount corresponds to the number of tokens you want transferred, and the destination is the hexadecimal public key of the receiving account. Your command should look like this

`npm run erc20iface transfer 200 0166795bb8895dcec5631690fa3d5dd3daacde7efeefb1e79176e9d879fd669b47`

> To send tokens from the address you just sent to, just change the `KEYS` constant to use the path of that accounts public and secret keys (in this case account `0166795bb8895dcec5631690fa3d5dd3daacde7efeefb1e79176e9d879fd669b47`).

Reviewer 3: It should be "account's" in "use the path of that accounts public and secret keys", since it's a possessive form.

> To send tokens from the address you just sent to, just change the `KEYS` constant to use the path of that accounts public and secret keys (in this case account `0166795bb8895dcec5631690fa3d5dd3daacde7efeefb1e79176e9d879fd669b47`).

Reviewer: ```suggestion
To send tokens from the address you just funded, just change the `KEYS` constant to use the path of that accounts public and secret keys (in this case account `0166795bb8895dcec5631690fa3d5dd3daacde7efeefb1e79176e9d879fd669b47`).
```

> 

Reviewer: Is there anything to say in conclusion? Maybe a short sentence to tie up the guide.
